File: 98.validate-binary-search-tree.cpp
// @leet start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <limits>

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (root == nullptr) {
            return true;
        }

        return isValidBSTHelper(root->left, std::numeric_limits<int64_t>::min(), root->val) &&
               isValidBSTHelper(root->right, root->val, std::numeric_limits<int64_t>::max());
    }

    bool isValidBSTHelper(TreeNode * node, int64_t min_val, int64_t max_val) {
        if (node == nullptr) {
            return true;
        }

        if (node->val <= min_val) {
            return false;
        }

        if (node->val >= max_val) {
            return false;
        }

        return isValidBSTHelper(node->left, min_val, node->val) &&
               isValidBSTHelper(node->right, node->val, max_val);
    }
};
// @leet end

File: 98.validate-binary-search-tree.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def validBSTHelper(node, curr_max, curr_min):
            if node == None:
                return True
            
            if node.val > curr_max or node.val < curr_min:
                return False

            return validBSTHelper(node.left, node.val, curr_min) and validBSTHelper(node.right, curr_max, node.val)

        return validBSTHelper(root, sys.maxsize, -sys.maxsize)
# @leet end
